src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -47,72 +52,137 @@ public static StreamingClientProvider getStreamingClientProviderInstance() {
+            })
+        .build(streamingClientHandler::createClient);
   }
 
+  /***************************** BEGIN SINGLETON CODE *****************************/
   private static final KCLogger LOGGER = new KCLogger(StreamingClientProvider.class.getName());
-  private SnowflakeStreamingIngestClient parameterEnabledClient;
+
   private StreamingClientHandler streamingClientHandler;
-  private Lock providerLock;
+  private LoadingCache<StreamingClientProperties, SnowflakeStreamingIngestClient> registeredClients;
 
-  // private constructor for singleton
+  /**
+   * Private constructor to retain singleton
+   *
+   * <p>If the one client optimization is enabled, this creates a threadsafe {@link LoadingCache} to
+   * register created clients based on the corresponding {@link StreamingClientProperties} built
+   * from the given connector configuration. The cache calls streamingClientHandler to create the
+   * client if the requested streaming client properties has not already been loaded into the cache.
+   * When a client is evicted, the cache will try closing the client, however it is best to still
+   * call close client manually as eviction is executed lazily
+   */
   private StreamingClientProvider() {
     this.streamingClientHandler = new StreamingClientHandler();
-    providerLock = new ReentrantLock(true);
+    this.registeredClients = buildLoadingCache(this.streamingClientHandler);
   }
 
   /**
    * Gets the current client or creates a new one from the given connector config. If client
    * optimization is not enabled, it will create a new streaming client and the caller is
-   * responsible for closing it
+   * responsible for closing it. If the optimization is enabled and the registered client is
+   * invalid, we will try recreating and reregistering the client
    *
    * @param connectorConfig The connector config
    * @return A streaming client
    */
   public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
+    SnowflakeStreamingIngestClient resultClient;
+    StreamingClientProperties clientProperties = new StreamingClientProperties(connectorConfig);
+
     if (Boolean.parseBoolean(
         connectorConfig.getOrDefault(
             SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
             Boolean.toString(ENABLE_STREAMING_CLIENT_OPTIMIZATION_DEFAULT)))) {
       LOGGER.info(
-          "Streaming client optimization is enabled, returning the existing streaming client if"
-              + " valid");
-      this.providerLock.lock();
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
```

Comment:
nit: should we move this log after the client creation is succeeded? Similar to the logic below when one client optimization is not enabled, then you can combine two log lines together.

Reply:
good point. Moved line to end of method so that we only have one log and added a warn log for when the registered client is invalid because ideally it should always be valid